src/compiler/crystal/syntax/parser.cr

```diff
@@ -3344,6 +3346,14 @@ module Crystal
           index += 1
         end
 
+        if name.ends_with?('=') && !name.ends_with?("[]=")
```

Comment:
The second test can be `name == "[]="`. There can't be anything before `[]=` in a method name.

However, is there a particular reason why `[]=` can be called like any method but all other setters can't? 
If not, the same rules should apply (with ` args.size != 2`).

Reply:
With `[]=` you can do `var[1, 2] = 3` to implement a matrix type for example

Reply:
Okay, that's a great use case. You can even do var`[1, foo: 2] = 3` 👀 

But it still doesn't explain `foo.[]=(foo, bar: bar) do |baz| end` 😃

Reply:
doh! I'll switch that to an equality check on `name`. @bew's totally right on the `[]=` with multi-arg case. There's at least one place in stdlib where `[]=` is given splat args and kwargs too.

Good catch on the block case for `[]=`. I'll address that.

Reply:
https://github.com/crystal-lang/crystal/blob/8d28cbb1fcd8fc462425f13b79203aa30f64bf96/src/compiler/crystal/syntax/parser.cr#L649-L653

Reply:
And because `(exp1, exp2)` is not a valid syntax, `f.x=(exp1, exp2)` is not a valid syntax either, so there's no way to pass multiple arguments there. It's the same as in Ruby (I copied it from Ruby, actually).

Reply:
And the exception for splat mentioned there maybe should be removed (it doesn't work in Ruby, so...)